src/main/java/com/sourcegraph/permission/PermissionRouter.java

```diff
@@ -0,0 +1,166 @@
+import com.atlassian.bitbucket.util.PageRequestImpl;
+import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
+import com.atlassian.sal.api.user.UserManager;
+import com.atlassian.sal.api.user.UserProfile;
+import com.sourcegraph.rest.Status;
+import org.roaringbitmap.RoaringBitmap;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+@Path("/permissions")
+@Component
+public class PermissionRouter {
+    @ComponentImport
+    private static RepositoryService repositoryService;
+    @ComponentImport
+    private static UserManager userManager;
+    @ComponentImport
+    private static UserService userService;
+    @ComponentImport
+    private static SecurityService securityService;
+
+    @Autowired
+    public PermissionRouter(RepositoryService repositoryService, UserManager userManager, UserService userService, SecurityService securityService) {
+        PermissionRouter.repositoryService = repositoryService;
+        PermissionRouter.userManager = userManager;
+        PermissionRouter.userService = userService;
+        PermissionRouter.securityService = securityService;
+    }
+
+    /**
+     * The getUsersWithRepositoryPermission endpoint returns a roaring bitmap containing the IDs of all the users
+     * that is granted the specified permission to the specified repository.
+     *
+     * Ex. /permissions/users?repository=PROJECT_1/rep_1&permission=read
+     */
+    @GET
+    @Path("/users")
+    public Response getUsersWithRepositoryPermission(@Context HttpServletRequest request, @QueryParam("repository") String repo, @QueryParam("permission") String perm) throws IOException {
+        UserProfile profile = userManager.getRemoteUser(request);
+        if (profile == null || !userManager.isSystemAdmin(profile.getUserKey())) {
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+        }
+
+        Repository repository = getRepository(repo);
+        if (repository == null) {
+            return Response.status(Response.Status.NOT_FOUND).entity("No such repository: " + repo).build();
+        }
+
+        Permission permission = getRepositoryPermission(perm);
+        if (permission == null) {
+            return Response.status(Status.UNPROCESSABLE_ENTITY).build();
+        }
+
+        RoaringBitmap bitmap = new RoaringBitmap();
+
+        UserSearchRequest.Builder builder = new UserSearchRequest.Builder();
+        builder.repositoryPermission(repository, permission);
+        UserSearchRequest search = builder.build();
+
+        PageRequest pageRequest = new PageRequestImpl(0, 5000);
```

Comment:
How "5000" gets chosen?

Reply:
It's an arbitrarily chosen value. I'm not exactly sure how it impacts memory as of yet. 

These were the endpoint response times for 10,000 repository IDs.
`100` page size -> 3.7s
`5000` page size -> 0.7s

Reply:
Gotcha, would you mind put what you replied as a block of comments in the code?